private/buf/bufgen/migrate.go

```diff
@@ -0,0 +1,553 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufgen
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/bufbuild/buf/private/buf/buffetch"
+	"github.com/bufbuild/buf/private/buf/bufgen/internal"
+	"github.com/bufbuild/buf/private/buf/bufgen/internal/bufgenplugin"
+	"github.com/bufbuild/buf/private/buf/bufgen/internal/bufgenv1"
+	"github.com/bufbuild/buf/private/buf/bufgen/internal/bufgenv2"
+	"github.com/bufbuild/buf/private/buf/bufref"
+	"github.com/bufbuild/buf/private/bufpkg/bufpluginexec"
+	"github.com/bufbuild/buf/private/pkg/encoding"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"go.uber.org/zap"
+)
+
+// Only do this in YAML because JSON does not allow comments.
+const migratedV2FileYAMLHeader = `# Generated by buf. Edit as necessary, and
+# remove this comment when you're finished.
+`
+
+type migrateV1ToV2Options struct {
+	input          string
+	genTemplate    string
+	types          []string
+	includePaths   []string
+	excludePaths   []string
+	includeImports bool
+	includeWKT     bool
+}
+
+func migrateV1ToV2(
+	ctx context.Context,
+	logger *zap.Logger,
+	readBucket storage.ReadBucket,
+	migrateOptions ...MigrateV1ToV2Option,
+) error {
+	options := &migrateV1ToV2Options{
+		input:       ".",
+		genTemplate: ExternalConfigFilePath,
+	}
+	for _, option := range migrateOptions {
+		option(options)
+	}
+	switch filepath.Ext(options.genTemplate) {
+	case ".json", ".yaml", ".yml":
+		// OK
+	default:
+		return errors.New("migration can only apply to a file on disk with extension .yaml, .yml or .json")
```

Comment:
Note: we are not migrating anything if the template is passed on command line, i.e. `buf generate --template {"version": "v1"...`
We don't know where to put the file if there is no file to begin with. Maybe print to stdout/stderr? But that goes against the direction of making the `buf generate` command shorter.